vies_service.go

```diff
@@ -0,0 +1,127 @@
+package vat
+
+import (
+	"bytes"
+	"encoding/xml"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// LookupServiceInterface is an interface for the service that calls external services to validate VATs.
+type LookupServiceInterface interface {
+	Validate(vatNumber string) error
+}
+
+// viesService validates EU VAT numbers with the VIES service
+type viesService struct{}
+
+// Validate returns whether the given VAT number is valid or not
+func (s *viesService) Validate(vatNumber string) error {
+	if len(vatNumber) < 3 {
+		return ErrInvalidVATNumberFormat
+	}
+
+	res, err := s.lookup(s.getEnvelope(vatNumber))
+	if err != nil {
+		return ErrServiceUnavailable{Err: err}
+	}
+	defer func() {
+		_ = res.Body.Close()
+	}()
+
+	xmlRes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return ErrUnexpected{Err: err}
```

Comment:
Again here, I'm thinking `ErrServiceUnavailable` or something along the lines like that. If we can't read the response then it's the same as saying the service is unavailable 🤔

Reply:
I guess I was assuming if `res` has a `Body`, which I think it must since it is not a pointer, then an error should never happen here. But I could be wrong here. Don't know enough about how all this reader stuff works in Go.

If you're fairly certain an error here indicates we did not get a good response and not that something else failed, then yeah I'd agree that `ErrServiceUnavailable` is the better option.

Reply:
We can define errors of every scenario here but thinking of how the caller of this stuff is handling errors from `vat` - if we don't get a readable response from VIES we might as well tell the caller that the service is unavailable as there's really no better way to describe the situation we're in.

If I were to imagine error handling for this, and say we had different errors for both `ErrBadVIESResponse` and `ErrServiceUnavailable`, I'd probably do something like:
```
err := vat.Validate(vatNumber)
if err != nil {
  if err == ErrServiceUnavailable || err == ErrBadVIESResponse {
    // do something about not being able to talk to VIES
  }
}
...
```
Error probably would be handled the same, whether VIES returned that it was unavailable or it returned garbage we can't parse.

That's my two cents anyway